docs/design/pnc.md

```diff
@@ -0,0 +1,113 @@
+**# Design document for PNC package manager
+
+Contents:
+
+1. [PNC](#PNC)
+2. [Current implementation overview (OSBS)](#overview-of-the-current-implementation-in-osbs)
+3. [Design for the Cachi2 implementation](#design-for-the-implementation-in-cachi2)
+
+## PNC
+
+### Glossary
+- **PNC**: Project NewCastle - [open-source project](https://github.com/project-ncl/pnc) for managing, executing, 
+  and tracking cross-platform builds.
+
+
+### Repository configuration 
+Projects that currently use OSBS to fetch sources from PNC do so through a file named `fetch-artifacts-pnc.yaml`
+([docs](https://osbs.readthedocs.io/en/osbs_ocp3/users.html?highlight=fetch#fetch-artifacts-pnc-yaml)).
+Example of a `fetch-artifacts-pnc.yaml` file:
+
+```yaml
+metadata:
+  # this object allows additional parameters, you can put any metadata here
+  author: shadowman
+builds:
+  # all artifacts are grouped by builds to keep track of their sources
+  - build_id: '1234' # build id must be string
+    artifacts:
+      # list of artifacts to fetch, artifacts are fetched from PNC using their IDs
+      - id : '12345' # artifact id must be string
+        # the target can just be a filename or path+filename
+        target: test/rhba-common-7.10.0.redhat-00004.pom
+      - id: '12346'
+        target: prod/rhba-common-7.10.0.redhat-00004-dist.zip
+  - build_id: '1235'
+    artifacts:
+      - id: '12354'
+        target: test/client-patcher-7.10.0.redhat-00004.jar
+      - id: '12355'
+        target: prod/rhdm-7.10.0.redhat-00004-update.zip
+```
+
+These artifacts are fetched to `artifacts/<target>` path at the root of the repository.
+
+### PNC instance configuration
+Currently, the url of the PNC instance is configured as a part of OSBS config. 
```

Comment:
This makes me wonder how are we going to implement integration tests for this. Is it possible to spin a basic PNC instance using something like `docker-compose`? All of our integration tests currently run on Github Actions, so we can't access a private instance.

Reply:
Looking at the PNC repositories, it does not look like it doesn't have dockerfiles/compose out of the box, and setting it all up could prove difficult. But since it's just an API, we could probably have a small mock flask application or something that would serve mock responses.

Reply:
You're right, but I wonder how much value an integration test with a mocked PNC would bring. Maybe we can consider some internal testing for this, then? I'm just afraid there could be cases we would not cover with a mocked API.

Reply:
To the best of my knowledge, we just need a couple of API responses, and a working url to download the files referenced in those responses. In my opinion, setting up an arbitrary PNC instance just for this is an overkill, but I could see adding some test artifacts to an existing internal PNC instance.

As for internal testing, could you please expand on how that would work? I understand that it would be e.g. a repo in gitlab with a  `fetch-artifacts-pnc.yaml` file poiting to some expected artifacts, but how would we tie that back in with the existing integration testing in this repository via github actions?

Reply:
> but how would we tie that back in with the existing integration testing in this repository via github actions?

One of the options would IMO be self-hosted runners: https://docs.github.com/en/actions/hosting-your-own-runners/managing-self-hosted-runners/about-self-hosted-runners
I don't have experience with github, but if it's anything like gitlab-runners then one just needs to set up an agent in the private environment and add it to the list of CI runners, in this case - actions. The beauty of at least gitlab runner is that it's not a push workflow, but a pull workflow, so even runners behind a firewall are able to report back provided the correct ports for egress are opened on the client side.
In practical terms, I can imagine using your already deployed instance for testing this way.

Reply:
@eskultety That is a really good idea! Who would own and maintain the self-hosted github runner? To the best of my knowledge and some quick looking around, we don't currently have such github runner instance available internally.

Reply:
> Who would own and maintain the self-hosted github runner?

From community POV that is usually the responsibility of people adding the support for a particular backend/driver as those are the people who actually need it :) ; but since the contributor may not always have direct involvement with the backend they're trying to add support for then the next obvious choice would IMO be whoever maintains the current internal deployment as they clearly understand the architecture the best and so it makes sense if a small infra adjustment is made such that a self hosted runner agent is created and wired so that it can relay jobs to the PNC deployment.

Reply:
We've had a meeting with @brunoapimentel today where we discussed this, and arrived at the conclusion that the best available option might be extensive unit testing. The self hosted github runner seem not viable from a maintenance perspective (I really doubt your team, my team or the PNC folks will have any interest in doing that) and also possibly a security perspective. 

We're still exploring if e2e konflux tests might be also viable for this, Bruno took a follow-up item on that. But, the plan at this moment is to proceed with extensive unit tests (as in testing the manager as a whole and mocking the API responses).